R/spq_filter.R

```diff
@@ -1,21 +1,82 @@
-#' Builds the "filter" part of a query.
+#' Filters results by adding conditions
 #' @inheritParams spq_arrange
 #' @export
 #' @examples
+#' spq_init() %>%
+#' spq_filter(item == wdt::P31("wd:Q13442814"))
 #' # Corpus of articles with "wikidata" in the title
 #' spq_init() %>%
 #' spq_add("?item wdt:P31 wd:Q13442814") %>%
 #' spq_add("?item rdfs:label ?itemTitle") %>%
 #' spq_filter(str_detect(str_to_lower(itemTitle), 'wikidata')) %>%
 #' spq_filter(lang(itemTitle)=="en") %>%
 #' spq_head(n = 5)
-spq_filter=function(query = NULL, ...){
-  variables = purrr::map_chr(rlang::enquos(...), spq_treat_argument)
+spq_filter = function(.query = NULL, ...){
+  filters = purrr::map(rlang::enquos(...), spq_treat_filter_argument)
 
-  query$filter <- c(
-    query$filter,
-    sprintf("FILTER(%s)", variables)
-  )
+  # FILTER filters :-)
+  normal_filters = filters[purrr::map_lgl(filters, is.character)]
+  if (length(normal_filters) > 0) {
+    .query$filter <- c(
+      .query$filter,
+      sprintf("FILTER(%s)", normal_filters)
+    )
+  }
+
+  # triple pattern "filters"
+  triple_filters <-  filters[purrr::map_lgl(filters, is.list)]
+  if (length(triple_filters) > 0) {
+    for (i in seq_along(triple_filters)) {
+      .query = spq_add(
+        .query,
+        .subject = triple_filters[[i]][["subject"]],
+        .verb = triple_filters[[i]][["verb"]],
+        .object = triple_filters[[i]][["object"]]
+      )
+    }
+  }
+
+  return(.query)
+}
+
+spq_treat_filter_argument = function(arg) {
+
+  eval_try = try(rlang::eval_tidy(arg), silent = TRUE)
+
+  if (is.spq(eval_try)) {
+    return(eval_try)
+  }
 
-  return(query)
+  code <- if (!inherits(eval_try, "try-error") && is.character(eval_try)) {
+    # e.g. `"desc(length)"`
+    eval_try
+  } else {
+    # e.g. `desc(length)`, without quotes
+    rlang::expr_text(arg) %>% stringr::str_replace("^~", "")
+  }
+
+  if (!grepl("::", code)) {
```

Comment:
I'm still not sure this is sufficient to differentiate the triple pattern filters and the "FILTER filters".

Reply:
#86

Reply:
It probably isn't, what if someone tries to use :: in the sense packagename::functionname() for a filter FILTER ? But, using such a function for a filter FILTER wouldn't work anyways, I guess?

Reply:
yes, exactly, you can't do that for a filter FILTER.

the case where this condition won't be enough is if someone used VALUES for all things with ::. So later, it'd be good to check there is either ::, or a string that is a VALUE for something with :. Does this make sense?

Reply:
I'm not sure I get what you mean. Do you mean people could list properties in VALUES so that no :: would appear directly in the call to spq_filter?

Reply:
Yes! Imagine something à la

```r
spq_init() %>%
  spq_set(blop = "wdt:thing", plouf = "wd:plouf") %>%
  spq_filter(item == blop(pouf))
```

The thing in `spq_filter()` has no `::`.

Reply:
Indeed... well, in that case, yes, probably it'd be good to check the occurrence of variable/pseudo-function names in the VALUES set.
(You're doing the questions and the answers really, oops)